### PR TITLE
Frontend validation

### PR DESCRIPTION
**Pull request by josh146**
* Adds frontend validation database, the `strawberryfields.devicespecs` submodule

* Defines the abstract base class `DeviceSpecs`, a pre-Python 3.7 'dataclass' for storing validation information for a backend. Note: after playing around with both Python 3.7 `dataclass` module and the Python 3+ `attrs` module, it seems to make more sense in the short term to simply take the approach used for device validation in PennyLane.

* Each supported backend by Strawberry Fields should have a class defined in the `data` submodule that inherits from `DeviceSpecs`, with required class attributes provided.

* `Program` class modified to validate against the new backend databases

### CODE REVIEW

File: strawberryfields/data/fock.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Fock backend validation data"""
+from .device import DeviceData
+
+
+class Fock(DeviceData):
```
Comment (smite) [inline strawberryfields/data/fock.py]: I'd name the DeviceData classes more descriptively, we have so many object names with 'Fock' in them. Maybe "FockData" or "FockCapabilities"?

Comment (josh146) [reply strawberryfields/data/fock.py]: `FockSpecs`?

File: strawberryfields/program.py [hunk 2]
```diff
@@ -772,49 +702,90 @@ def compile(self, backend='fock', optimize=True):
         Returns:
             Program: compiled program
         """
-        db = backend_database[backend]
+        # TODO: Avoid circular import
+        from strawberryfields.data import backend_databases
```
Comment (smite) [inline strawberryfields/program.py]: I think circular imports can be avoided if you use "import X" instead of "from X import Y".

Comment (josh146) [reply strawberryfields/program.py]: I tried `import strawberryfields.data` but no luck. From what I can tell, the issue is that `data.chip0` imports `io`, which then imports `Program`.

File: strawberryfields/data/fock.py (hunk 1 above)
Comment (smite) [reply strawberryfields/data/fock.py]: Not bad, maybe also rename the base class to `DeviceSpecs` for consistency?

Comment (josh146) [reply strawberryfields/data/fock.py]: Yep. We could also update the module to `strawberryfields.devicespecs`, for consistency, using `data` feels too general

File: strawberryfields/program.py (hunk 2 above)
Comment (josh146) [reply strawberryfields/program.py]: Fixed in the latest commit

File: strawberryfields/devicespecs/device_specs.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+            int: number of supported modes
+        """
+
+    @property
+    @abc.abstractmethod
+    def remote(self) -> bool:
+        """Whether the backend supports remote execution.
+
+        Returns:
+            bool: returns ``True`` if the backend supports remote execution.
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]:
```suggestion
            bool: ``True`` if the backend supports remote execution
```

File: strawberryfields/devicespecs/device_specs.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        """
+
+    @property
+    @abc.abstractmethod
+    def interactive(self) -> bool:
+        """Whether the backend can be used interactively, that is,
+        the backend state is not reset between engine executions.
+
+        Returns:
+            bool: returns ``True`` if the backend supports interactive use.
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]:
```suggestion
            bool: ``True`` if the backend supports interactive use
```

File: strawberryfields/devicespecs/device_specs.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+            {'operation_name': {'option1': val, 'option2': val,...}}
+
+        For each operation specified in the dictionary, the
+        :meth:`~Operation.decompose` method will be called during
+        :class:`Program` compilation, with keyword arguments
+        given by the dictionary value.
+
+        Returns:
+            dict[str, dict]: the quamtum operations that are supported
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]:
```suggestion
            dict[str, dict]: the quantum operations that are supported
```

File: strawberryfields/devicespecs/device_specs.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+    @property
+    def parameter_ranges(self) -> Dict[str, List[List[float]]]:
+        """Allowed parameter ranges for supported quantum operations.
+
+        This property is optional.
+
+        Returns:
+            dict[str, list]: a dictionary mapping an allowed quantum operation
+            to a nested list of the form ``[[p0_min, p0_max], [p1_min, p0_max], ...]``.
+            where ``pi`` corresponds to the ``i``th gate parameter.z
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]:
```suggestion
            where ``pi`` corresponds to the ``i``th gate parameter.
```

File: strawberryfields/devicespecs/device_specs.py [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+        """
+        return dict()
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]: How come this one returns something? :thinking:

File: strawberryfields/devicespecs/device_specs.py [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    @property
+    def topology(self):
+        """The allowed circuit topology of the backend device as a directed
+        acyclic graph.
+
+        This property is optional; if arbitrary topologies are allowed by the device,
+        this will simply return ``None``.
+
+        Returns:
+            networkx.DAGGraph: a directed acyclic
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]:
```suggestion
            networkx.DAGGraph: a directed acyclic graph
```

File: strawberryfields/devicespecs/device_specs.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        """
+        if self.blackbird_template is None:
+            return None
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]: To confirm logic here: non-existance of `blackbird_template` implies that arbitrary topologies are allowed? Is it better to make this explicit? i.e., have a property like `all_topologies` or something?

File: strawberryfields/devicespecs/device_specs.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+        # returned DAG has all parameters set to 0
+        bb = blackbird.loads(self.blackbird_template)
+
+        if bb.is_template():
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]: isn't it confusing that something found under `self.blackbird_template` isn't guaranteed to be a template?

File: strawberryfields/devicespecs/device_specs.py [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+    @property
+    def blackbird_template(self):
+        """The Blackbird template that will be accepted by the backend device.
+
+        This property is optional; if arbitrary topologies are allowed by the device,
+        this will simply return ``None``.
+
+        If the device expects a specific template for the recieved Blackbird
+        script, this method will return the seralized Blackbird template in string
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]:
```suggestion
        script, this method will return the serialized Blackbird template in string
```

File: strawberryfields/devicespecs/device_specs.py [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+        form.
+
+        Returns:
+            str: Blackbird template
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]: want to also put that `None` is a valid return type here?

File: strawberryfields/ops.py [hunk 13]
```diff
@@ -1814,30 +1816,24 @@ def __init__(self, V, r=None, decomp=True, tol=1e-6):
...
-            raise NotImplementedError
-
         p = _unwrap(self.p)
         s = sqrt(sf.hbar / 2)  # scaling factor, since the backend API call is hbar-independent
         backend.prepare_gaussian_state(p[1]/s, p[0]/(s*s), reg)
 
     def _decompose(self, reg):
         # pylint: disable=too-many-branches
-        if not self.decomp:
-            return None  # refuse to decompose
```
Comment (co9olguy) [inline strawberryfields/ops.py]: Why was this here before?

File: strawberryfields/program.py [hunk 14]
```diff
@@ -765,49 +696,99 @@ def compile(self, backend='fock', **kwargs):
         Returns:
             Program: compiled program
         """
-        db = backend_database[backend]
+        if backend in specs.backend_databases:
```
Comment (co9olguy) [inline strawberryfields/program.py]: do we want to go with `database`? might confuse people that we have a relational DB or something?

File: strawberryfields/program.py [hunk 15]
```diff
@@ -765,49 +696,99 @@ def compile(self, backend='fock', **kwargs):
...
+                    # by passing the kwarg `decomp=False` to more
+                    # operations (currently only ops.Gaussian allows this).
+                    #
+                    # For example, the 'gaussian' backend supports setting the state
+                    # via passing directly the (mu, cov) OR by first having the
+                    # frontend decompose into other primitive gates.
+                    # That is, ops.Gaussian is both a primitive _and_ a decomposition
+                    # for the 'gaussian' backend, and it's behaviour can be chosen
+                    # by the user.
+                    if (op_name in db.primitives) and hasattr(cmd.op, 'decomp'):
```
Comment (co9olguy) [inline strawberryfields/program.py]: Just want to follow logic here: this block just checks the case that an op might be both decomposable and also a primitive, and confirms user-override before performing decomp?

File: strawberryfields/program.py [hunk 16]
```diff
@@ -765,49 +696,99 @@ def compile(self, backend='fock', **kwargs):
...
+            # check topology
+            grid = self._list_to_grid(seq)
+            DAG = self._grid_to_DAG(grid)
+
+            # relabel the DAG nodes to integers, with attributes
+            # specifying the operation name. This allows them to be
+            # compared, rather than using Command objects.
+            mapping = {i: n.op.__class__.__name__ for i, n in enumerate(DAG.nodes())}
+            circuit = nx.convert_node_labels_to_integers(DAG)
+            nx.set_node_attributes(circuit, mapping, name='name')
```
Comment (co9olguy) [inline strawberryfields/program.py]: Getting a bit confused. Does 'name' refer to the generic name of the Command (e.g., 'BSgate') or to a unique name associated with each node?

File: strawberryfields/devicespecs/chip0.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    # mapping for the interferometer when #87 is merged
+    decompositions = {"Interferometer": {}}
+
+    # TODO: update the below to specify the rectangular_symmetric
+    # mapping for the interferometer when #87 is merged.
+    # The current topology defined below is just for demonstration
+    blackbird_template = textwrap.dedent(
+        """\
+        name chip0_template
+        version 1.0
```
Comment (co9olguy) [inline strawberryfields/devicespecs/chip0.py]: Can we just set everything throughout the library to the placeholder `version 0.0` until everything is more stable?

File: strawberryfields/devicespecs/chip0.py [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+        target chip0 (shots=1)
+
+        # for n spatial degrees, first n signal modes, then n idler modes, phase zero
```
Comment (co9olguy) [inline strawberryfields/devicespecs/chip0.py]: This comment, and earlier `modes` variable indicates this `devicespec` supports an arbitrary number of modes. However, the actual circuit has 4 modes.

File: strawberryfields/devicespecs/fock.py [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+        "Rgate",
+        "Vgate",
+        "Kgate",
+        "Fouriergate",
+        "BSgate",
+        "CKgate",
+    }
+
+    decompositions = {
+        "Interferometer": {},
```
Comment (co9olguy) [inline strawberryfields/devicespecs/fock.py]: Can you remind me what effect a blank dictionary has? Does that mean it supports any possible decomposition the frontend might try?

File: strawberryfields/devicespecs/device_specs.py [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+
+        If the device supports arbitrary number of modes, set this to 0.
+
+        Returns:
+            int: number of supported modes
+        """
+
+    @property
+    @abc.abstractmethod
+    def remote(self) -> bool:
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]: I suppose we also want to have a `local` property as well then, since various combinations or remote/local execution are possible depending on the device

File: tests/frontend/test_devicespecs.py [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+            the abstract base class"""
+            modes = 0
+            remote = False
+            interactive = True
+            primitives = set()
+            decompositions = set()
+
+        dummy = DummyDevice()
+
+        assert not dummy.parameter_ranges
```
Comment (co9olguy) [inline tests/frontend/test_devicespecs.py]: Per docstring, do we also want to test that it is a dictionary that is returned?

File: tests/frontend/test_devicespecs.py [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+        # check gates are correct
+        assert circuit[0]['name'] == 'Sgate'
+        assert circuit[1]['name'] == 'Dgate'
+        assert circuit[2]['name'] == 'BSgate'
+        assert circuit[3]['name'] == 'MeasureFock'
+        assert circuit[4]['name'] == 'MeasureFock'
+
+        # check arguments
+        assert circuit[0]['args'] == [0, 0]
+        assert circuit[1]['args'] == [-7.123]
```
Comment (co9olguy) [inline tests/frontend/test_devicespecs.py]: Just curious why the Dgate parameter does not get converted to zero?

File: tests/frontend/test_program.py [hunk 23]
```diff
@@ -305,3 +309,269 @@ def test_merge_incompatible(self):
...
+        prog = sf.Program(3)
+        r = 0.453
+        cov = np.array([[np.exp(-2*r), 0], [0, np.exp(2*r)]])*sf.hbar/2
+
+        with prog.context:
+            ops.Gaussian(cov, decomp=True) | 0
+
+        new_prog = prog.compile(backend='gaussian')
+
+        print([i.op.__class__.__name__ for i in new_prog.circuit])
```
Comment (co9olguy) [inline tests/frontend/test_program.py]:
```suggestion
```

File: tests/frontend/test_program.py [hunk 24]
```diff
@@ -305,3 +309,269 @@ def test_merge_incompatible(self):
...
+        assert len(new_prog) == 1
+
+        # decomposed gate should be an Sgate
+        circuit = new_prog.circuit
+        assert circuit[0].op.__class__.__name__ == "Sgate"
+        assert circuit[0].op.p[0] == r
+
+    def test_topology_validation(self, monkeypatch):
+        """Test compilation properly matches the device toplogy"""
```
Comment (co9olguy) [inline tests/frontend/test_program.py]:
```suggestion
        """Test compilation properly matches the device topology"""
```

File: strawberryfields/devicespecs/device_specs.py [hunk 25]
```diff
@@ -0,0 +1,144 @@
...
+"""Abstract base class for storing device data for validation"""
+from typing import List, Set, Dict
+import abc
+
+import blackbird
+from blackbird.utils import to_DiGraph
+
+
+class DeviceSpecs(abc.ABC):
+    """Abstract base class for backend data"""
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: :+1: I'll extend this with proper documentation

File: strawberryfields/devicespecs/__init__.py [hunk 26]
```diff
@@ -0,0 +1,29 @@
...
+# limitations under the License.
+"""Device validation submodule"""
+from .chip0 import Chip0Specs
+from .fock import FockSpecs
+from .gaussian import GaussianSpecs
+from .tensorflow import TFSpecs
+
+
+backend_databases = {
+    "base": FockSpecs,
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/devicespecs/__init__.py]: Yeah, I wasn't sure what to do for the `BaseBackend`? I could just make another dataclass for the base backend, that pretends to support all gates

File: strawberryfields/devicespecs/device_specs.py [hunk 27]
```diff
@@ -0,0 +1,144 @@
...
+
+    @property
+    @abc.abstractmethod
+    def modes(self) -> int:
+        """The supported number of modes of the device.
+
+        If the device supports arbitrary number of modes, set this to 0.
```
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: I deliberated over this and eventually found a reason why `None` wouldn't work (I had https://github.com/XanaduAI/pennylane/issues/178 in the back of my mind). But for the life of me, I can't remember why :thinking: I'll look back into this.

File: strawberryfields/devicespecs/device_specs.py (hunk 5 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: lol oops

File: strawberryfields/devicespecs/device_specs.py (hunk 7 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Should it return `None`? I figured that

1. This shouldn't be an abstract base class, as only a very small subset of devices will need to set this, and

2. Better to be consistent, and just return an empty dictionary instead of `None`. It also leads to clearer code, i.e., `if db.parameter_range:` as opposed to `if db.parameter_range is not None:`

File: strawberryfields/devicespecs/device_specs.py (hunk 10 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Yeah, I don't like the name `blackbird_template` for the property anyway. Some suggestions:

* `DeviceSpec.topology` for the blackbird script

* `DeviceSpec.graph` for the DAG

File: strawberryfields/ops.py (hunk 13 above)
Comment (josh146) [reply strawberryfields/ops.py]: This was required to 'hack' around the previous try-except logic for decompositions. Now that that has been removed, it is no longer needed.

File: strawberryfields/program.py (hunk 14 above)
Comment (josh146) [reply strawberryfields/program.py]: How about `backend_specs`?

File: strawberryfields/program.py (hunk 15 above)
Comment (josh146) [reply strawberryfields/program.py]: Yep!

File: strawberryfields/program.py (hunk 16 above)
Comment (josh146) [reply strawberryfields/program.py]: You're correct, name refers to the operation name `BSgate`.

File: strawberryfields/devicespecs/fock.py (hunk 19 above)
Comment (josh146) [reply strawberryfields/devicespecs/fock.py]: Essentially, yes. So for instance, once #87 is merged, the `chip0` device will look like this:
```python
   decompositions = {
        "Interferometer": {"geometry": "mach_zehnder"},
   }
```
specifying that compilation of the interferometer will use the `mach_zehnder` decomposition function, as opposed to `rectangular` or `triangular`.

File: strawberryfields/devicespecs/device_specs.py (hunk 20 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Ah, I didn't think of this. How would this look? I assumed `remote` and `local` would be mutually exclusive

File: strawberryfields/devicespecs/device_specs.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Changed to `None` for arbitrary number of modes.

File: strawberryfields/devicespecs/device_specs.py (hunk 9 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Since this these two class attributes are mutually exclusive, it might be better to not allow both to be defined?

File: strawberryfields/devicespecs/device_specs.py (hunk 10 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Made the change above :+1:

File: strawberryfields/program.py (hunk 14 above)
Comment (josh146) [reply strawberryfields/program.py]: Done

File: strawberryfields/devicespecs/device_specs.py (hunk 12 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Done

File: strawberryfields/devicespecs/chip0.py (hunk 17 above)
Comment (josh146) [reply strawberryfields/devicespecs/chip0.py]: Done

File: strawberryfields/devicespecs/chip0.py (hunk 18 above)
Comment (josh146) [reply strawberryfields/devicespecs/chip0.py]: fixed

File: tests/frontend/test_devicespecs.py (hunk 21 above)
Comment (josh146) [reply tests/frontend/test_devicespecs.py]: Done

File: tests/frontend/test_devicespecs.py (hunk 22 above)
Comment (josh146) [reply tests/frontend/test_devicespecs.py]: Because it is hardcoded into the circuit attribute:
```python
circuit = """
    name test
    version 0.0

    Sgate({sq}, 0) | 0
    Dgate(-7.123) | 1
    BSgate({theta}) | 0, 1
    MeasureFock() | 0
    MeasureFock() | 2
    """
```
This takes into account devices that potentially have fixed gate parameters, i.e., fixed squeezing.

File: strawberryfields/devicespecs/device_specs.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: done

File: strawberryfields/devicespecs/__init__.py [hunk 28]
```diff
@@ -11,14 +11,58 @@
...
+
+This module implements the :class:`~.DeviceSpecs` class, an abstract base
+data class used to store details and specifications of the Strawberry Fields
+backend and devices.
+
+These details are used by the :class:`~.Program` class when validating and
+compiling quantum programs. By querying the data class corresponding to the
+requested device/backend, the :class:`~.Program` will be able to:
+
+1. **Validate** that the program has the correct number of modes, valid
```
Comment (co9olguy) [inline strawberryfields/devicespecs/__init__.py]: "and consists of valid quantum operations for that device"

Comment (co9olguy) [reply strawberryfields/devicespecs/__init__.py]: sorry, can't suggest

File: strawberryfields/devicespecs/__init__.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (co9olguy) [reply strawberryfields/devicespecs/__init__.py]: I think that is preferable to puttingin the wrong specs. Alternatively, you could just look at which gates are explicitly covered by BaseBackend (even though it doesn't do any simulation)

File: strawberryfields/devicespecs/device_specs.py (hunk 7 above)
Comment (co9olguy) [reply strawberryfields/devicespecs/device_specs.py]: thumb

File: strawberryfields/devicespecs/device_specs.py (hunk 9 above)
Comment (co9olguy) [reply strawberryfields/devicespecs/device_specs.py]: Sure, as long as they are mutually exclusive. It just wasn't obvious to me at first glance that that was the case

File: strawberryfields/devicespecs/device_specs.py [hunk 29]
```diff
@@ -93,26 +93,26 @@ def parameter_ranges(self) -> Dict[str, List[List[float]]]:
...
 
         This property is optional; if arbitrary topologies are allowed by the device,
         this will simply return ``None``.
 
         Returns:
-            networkx.DAGGraph: a directed acyclic
+            networkx.DiGraph: a directed acyclic graph
         """
-        if self.blackbird_template is None:
+        if self.circuit is None:
```
Comment (co9olguy) [inline strawberryfields/devicespecs/device_specs.py]: I didn't catch where `self.circuit` is actually declared. Did I miss something?

File: strawberryfields/program.py [hunk 30]
```diff
@@ -696,11 +696,16 @@ def compile(self, backend='fock', **kwargs):
...
-            db = specs.backend_databases[backend]()
+        if backend in specs.backend_specs:
+            db = specs.backend_specs[backend]()
         else:
             raise ValueError("Could not find backend {} in Strawberry Fields database".format(backend))
 
+        if db.modes is not None:
+            if self.num_subsystems > db.modes:
+                raise CircuitError("This program requires {} modes, but the {} backend "
+                                   "only supports an {} mode program".format(self.num_subsystems, backend, db.modes))
```
Comment (co9olguy) [inline strawberryfields/program.py]: an->a

File: strawberryfields/devicespecs/device_specs.py (hunk 9 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Okay, I'll make this more clear

File: strawberryfields/program.py (hunk 30 above)
Comment (josh146) [reply strawberryfields/program.py]: done

File: strawberryfields/devicespecs/__init__.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (josh146) [reply strawberryfields/devicespecs/__init__.py]: Done (did as you suggested, made `BaseSpecs` which simply allows what's covered by `BaseBackend`).

File: strawberryfields/program.py (hunk 16 above)
Comment (co9olguy) [reply strawberryfields/program.py]: Ok, so this code checks that the graph topology matches, and that at each node, the ops have the same name. Does it check that these ops have the same parameters?

File: strawberryfields/devicespecs/device_specs.py (hunk 20 above)
Comment (co9olguy) [reply strawberryfields/devicespecs/device_specs.py]: There are two options:

- Have `DeviceSpec`s registered separately for local and remote versions of backends that are otherwise the same (e.g., Gaussian backend running locally or remotely). 

- Have a single `DeviceSpec` for a backend type, and extra properties `local` and `remote` which indicate what operation types are supported (could be one or the other, or both).

I think second is much easier

File: tests/frontend/test_devicespecs.py (hunk 22 above)
Comment (co9olguy) [reply tests/frontend/test_devicespecs.py]: ok so the parameters that are programmable should all get converted to zeros, but others will stay as specified?

File: strawberryfields/devicespecs/device_specs.py (hunk 9 above)
Comment (co9olguy) [reply strawberryfields/devicespecs/device_specs.py]: I don't see any changes. Was this made clearer in another location/file?

File: tests/frontend/test_devicespecs.py (hunk 22 above)
Comment (josh146) [reply tests/frontend/test_devicespecs.py]: Yes. I'm not sure I like this (using zero/non-zero to distinguish free/fixed parameters), but it needs more thinking.

File: strawberryfields/devicespecs/device_specs.py (hunk 9 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Line 134 and below:

https://github.com/XanaduAI/strawberryfields/blob/0e010d373f11866f8321a385e097b78de47ba4dc/strawberryfields/devicespecs/device_specs.py#L134

File: strawberryfields/program.py (hunk 16 above)
Comment (josh146) [reply strawberryfields/program.py]: Not yet. Currently, it just checks the name.

I tried to code this in, but it sort of bleeds into the compilation problem as well, so I removed that code. Do you want me to add it back/finish it?

Comment (co9olguy) [reply strawberryfields/program.py]: So currently two graphs are considered isomorphic if they have all the same gates in the same places, but whether arguments match or not does not influence it?

File: strawberryfields/devicespecs/device_specs.py (hunk 20 above)
Comment (josh146) [reply strawberryfields/devicespecs/device_specs.py]: Added the second option :+1:

File: strawberryfields/program.py (hunk 16 above)
Comment (josh146) [reply strawberryfields/program.py]: Yes, that is not currently done - this will be done in a new ticket.
